fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/GraphExporter.java

```diff
@@ -177,67 +208,120 @@ protected String getVertexID(final N node) {
+        for (final N n : network.nodes()) {
+            renderNode(context,
+                    n,
+                    vertexAttributeProvider.apply(n));
+        }
+    }
 
     /**
      * Render a node. To be implemented by subclass.
      *
-     * @param out the writer
+     * @param context the context
      * @param node the node to be rendered
      * @param attributes the attributes of the node
      */
-    protected abstract void renderNode(PrintWriter out,
+    protected abstract void renderNode(ExporterContext context,
                                        N node,
-                                       Map<String, String> attributes);
+                                       Map<String, Object> attributes);
+
+    /**
+     * Render all edges in a given network.
+     * @param context the context to use
+     */
+    protected void renderEdges(final ExporterContext context) {
+        final ImmutableNetwork<N, E> network = context.getNetwork();
+
+        // edge set
+        for (final E e : network.edges()) {
+            final EndpointPair<N> endpointPair = network.incidentNodes(e);
+            final N u = endpointPair.nodeU();
+            final N v = endpointPair.nodeV();
+            renderEdge(context,
+                    network.isDirected(),
+                    u,
+                    v,
+                    edgeAttributeProvider.apply(e));
+        }
+    }
 
     /**
      * Render an edge. To be implemented by subclass.
      *
-     * @param out the writer
+     * @param context the context
      * @param isDirected true iff edge is directed
-     * @param edge the edge to be rendered
      * @param source the source node of the edge
      * @param target the target node of the edge
      * @param attributes the attributes of the edge
      */
-    protected abstract void renderEdge(PrintWriter out,
+    protected abstract void renderEdge(ExporterContext context,
                                        boolean isDirected,
-                                       E edge,
                                        N source,
                                        N target,
-                                       Map<String, String> attributes);
+                                       Map<String, Object> attributes);
+    
+    /**
+     * Render all sub clusters in a given network.
+     * @param context the context to use
+     */
+    protected void renderClusters(final ExporterContext context) {
+        final ImmutableNetwork<N, E> network = context.getNetwork();
+        // render clusters
+        final Map<N, Set<N>> clusterMap = clusterProvider.apply(network);
+        int i = 1;
+        for (final Entry<N, Set<N>> cluster : clusterMap.entrySet()) {
+            renderCluster(context,
+                    String.valueOf(i),
+                    cluster.getKey(),
+                    cluster.getValue(),
+                    clusterAttributeProvider.apply(cluster.getKey()));
+            i ++;
```

Comment:
```suggestion
            i++;
```

Reply:
Wait, unary operators have no space in between?